src/receiver.rs

```diff
@@ -1,78 +1,344 @@
+                if any_flag(flags, S_CLOSE) {
+                    Ok(unsafe { self.recover_unchecked().unwrap() })
+                } else {
+                    Err(RecoverError::Live)
+                }
+            }
         })
     }
 
-    /// Attempts to receive. On failure, if the channel is not closed,
-    /// returns self to try again.
-    #[inline]
-    pub fn try_recv(mut self) -> Result<T, TryRecvError<T>> {
-        let state = self.inner.state();
-        match self.handle_state(state) {
-            Poll::Ready(Ok(x)) => Ok(x),
-            Poll::Ready(Err(Closed())) => Err(TryRecvError::Closed),
-            Poll::Pending => Err(TryRecvError::Empty(self)),
-        }
+    /// Returns a new [`Sender`] without checking the old one has closed.
+    ///
+    /// ## Safety
+    ///
+    /// You must not permit multiple live Senders to exist.
+    pub unsafe fn recover_unchecked(&mut self) -> Result<sender::Sender<'a, T>, Closed> {
+        self.hatch.map(|h| {
+            h.recycle(); // A release store
+            self.flags &= !LONELY; // reset flag
+            sender::Sender::new(h)
+        }).ok_or(Closed)
     }
+
+    /// Gets the value of the `mark_on_drop` option. See the [`Receiver`] docs for an explanation.
+    pub fn get_mark_on_drop(&self) -> bool { any_flag(self.flags, MARK_ON_DROP) }
+
+    /// Sets the `mark_on_drop` option. See the [`Receiver`] docs for an explanation.
+    pub fn mark_on_drop(mut self, on: bool) -> Self {
+        self.flags = toggle_flag(self.flags, MARK_ON_DROP, on);
+        self
+    }
+
+    /// Gets the value of the `close_on_receive` option. the [`Receiver`] docs for an explanation.
+    pub fn get_close_on_receive(&self) -> bool { any_flag(self.flags, CLOSE_ON_SUCCESS) }
+
+    /// Sets the `close_on_receive` option. See the [`Receiver`] docs for an explanation.
+    pub fn close_on_receive(mut self, on: bool) -> Self {
+        self.flags = toggle_flag(self.flags, CLOSE_ON_SUCCESS, on);
+        self
+    }
+
+    /// Drops our reference to the Hatch without any synchronisation.
+    /// Useful with external synchronisation to avoid a lock cycle.
+    ///
+    /// Correct use of this generally implies both the Sender and
+    /// Receiver leaking and some external management of the Hatch.
+    ///
+    /// # Safety
+    ///
+    /// This isn't technically unsafe, but...
+    /// * It could cause a memory leak.
+    /// * It could cause the Sender to wait forever.
+    pub unsafe fn leak(mut receiver: Receiver<T>) { receiver.hatch.take(); }
 }
 
-impl<T> Future for Receiver<T> {
-    type Output = Result<T, Closed>;
-    fn poll(self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Result<T, Closed>> {
-        let this = Pin::into_inner(self);
-        match this.handle_state(this.inner.state()) {
-            Poll::Pending => {},
-            x => return x,
+impl<'a, T> Drop for Receiver<'a, T> {
+    fn drop(&mut self) {
+        if let Some(hatch) = self.hatch.take() {
```

Comment:
Do we really require the `Option` wrapping for `self.hatch`? It seems this is only required for `hatch.cleanup` to be able to move the `Holder` object into it. That method is unsafe anyway and can't be called always even when we own a `Hatch`.

Reply:
The option wrapping is so we can "disable" the destructor.